src/Microsoft.OData.Edm/ExtensionMethods/TargetPathHelper.cs

```diff
@@ -0,0 +1,287 @@
+            }
+
+            IEdmType edmType = null;
+
+            if (entityContainerElement is IEdmEntitySet edmEntitySet)
+            {
+                edmType = edmEntitySet.Type.AsElementType();
+            }
+
+            if (entityContainerElement is IEdmSingleton edmSingleton)
+            {
+                edmType = edmSingleton.Type.AsElementType();
+            }
+
+            if (edmType is IEdmStructuredType structuredType)
+            {
+                // .../MyEntitySet/MyProperty/...
+                // .../MyEntitySet/MyNavigationProperty/...
+                // .../MySingleton/MyProperty/...
+                // .../MySingleton/MyNavigationProperty/...
+
+                return HandleStructuredType(model, pathSegments, structuredType, targetSegments, index, caseInsensitive);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns a property from a schema type.
+        /// </summary>
+        /// <param name="model">The edm model.</param>
+        /// <param name="pathSegments">The path segments of already resolved target segments.</param>
+        /// <param name="schemaType">The schema type.</param>
+        /// <param name="targetSegments">The target segments to resolve.</param>
+        /// <param name="index">The index of the target segments.</param>
+        /// <param name="caseInsensitive">Property name case-insensitive or not.</param>
+        /// <returns>An <see cref="IEdmVocabularyAnnotatable"/> object.</returns>
+        private static IEdmVocabularyAnnotatable HandleSchemaType(IEdmModel model, List<IEdmElement> pathSegments, IEdmSchemaType schemaType, string[] targetSegments, int index, bool caseInsensitive)
+        {
+            // Validate schema type is not followed by schema type e.g entitycontainer/entityset/NS.TypeCase1/NS.TypeCase2/...
+            IEdmSchemaType nextSchemaType = model.FindType(targetSegments[index]);
+
+            if (nextSchemaType != null)
+            {
+                throw new InvalidOperationException(Strings.TypeCast_HierarchyNotFollowed(schemaType, nextSchemaType));
+            }
+
+            if (schemaType is IEdmStructuredType structuredType)
+            {
+                return HandleStructuredType(model, pathSegments, structuredType, targetSegments, index, caseInsensitive);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Validate a schema type.
+        /// </summary>
+        /// <param name="schemaType">The schema type.</param>
+        /// <param name="pathSegments">The path segments of already resolved target segments.</param>
+        /// <param name="index">The index of the path segments.</param>
+        private static void ValidateSchemaType(IEdmSchemaType schemaType, List<IEdmElement> pathSegments, int index)
+        {
+            IEdmElement previousElement = pathSegments[index - 1];
+
+            bool isValid = IsDerivedTypeValid(previousElement, schemaType);

```

Comment:
Is it not misleading to name this "IsDerivedTypeValid" given that in that method you use "IsOrInheritsFrom"? In addition, `Strings.TypeCast_HierarchyNotFollowed` suggests that you're only handling a cast segment